### ISSUE TITLE
CustomPainter's paint method is not getting called before WidgetsBinding.instance.addPostFrameCallback in case of Multiple navigation

### ISSUE THREAD

**Issue by ratedRA** [flutter/flutter#118164]
I have a Flutter StatefulWidget and in initState() method I am using WidgetsBinding.instance.addPostFrameCallback to use one instance variable (late List _tracks). like -
`WidgetsBinding.instance.addPostFrameCallback((_) {
      for(itr = 0; itr<_tracks.length; itr++){
         // some logic
      }
    });`

As this would get invoked after all Widgets are done. In one of the CustomPaint's painter class I am initializing that variable.
`SizedBox.expand(
       child: CustomPaint(
            painter: TrackPainter(
                   trackCalculationListener: (tracks) {
                     _tracks = tracks;
              }),
           ),
        ),`

It is working fine when I have one screen, i.e the same class. But, When I am adding one screen before that and trying to navigate to this screen from the new screen it is throwing _tracks is not initialized exception.

In single screen case the paint method of painter is getting called before postFrameCallback but in case of multiple it is not getting before postFrameCallback and because of that the variable is not getting initialized.

**Comment by exaby73** [comment]
Hello @ratedRA. Thank you for filing this issue. Can you please provide the following information?

- A [complete and minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Ideally, a single `main.dart` file that can be run directly.
- The output of `flutter doctor -v`

Please provide the information in the form of text. See how to make collapsible sections with Markdown [here](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections).

**Comment by exaby73** [comment]
Hello @ratedRA. Your code while complete, is not minimal. Please read [complete and minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) on how to produce a minimal sample

**Comment by huycozy** [comment]
@ratedRA Please try this below sample code and check if it's applicable to you:

<details>
<summary>Sample code (updated)</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      routes: {
        '/mainMenu': (context) => const MainMenu(),
        '/game': (context) => const MyHomePage(title: 'game'),
      },
      initialRoute: '/mainMenu',
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final ValueNotifier<List<Rect>> _playerTracks = ValueNotifier([]);

  @override
  void initState() {
    super.initState();

    _playerTracks.addListener(() {
      print('value changed');
      // continue your logic here
      _playerTracks.value.forEach((element) {
        print(element.toString());
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: SingleChildScrollView(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              Container(
                color: Colors.white,
                margin: const EdgeInsets.all(20),
                child: AspectRatio(
                  aspectRatio: 1,
                  child: SizedBox.expand(
                    child: CustomPaint(
                      painter: RectanglePainter(
                          trackCalculationListener: (playerTracks) {
                            _playerTracks.value = playerTracks;
                          }),
                    ),
                  ),
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}

class MainMenu extends StatefulWidget {
  static String route = '/mainMenu';
  const MainMenu({super.key});

  @override
  State<MainMenu> createState() => _MainMenuState();
}

class _MainMenuState extends State<MainMenu> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Container(
          height: 200.0,
          color: Colors.white,
          child: ElevatedButton(
            onPressed: () {
              Navigator.pushNamed(context, '/game');
            },
            child: const Text('play game'),
          ),
        ),
      ),
    );
  }
}

class RectanglePainter extends CustomPainter {
  Function(List<Rect>) trackCalculationListener;

  RectanglePainter({required this.trackCalculationListener});

  @override
  void paint(Canvas canvas, Size size) {
    final Rect rect = Offset.zero & size;
    const RadialGradient gradient = RadialGradient(
      center: Alignment(0.7, -0.6),
      radius: 0.2,
      colors: <Color>[Color(0xFFFFFF00), Color(0xFF0099FF)],
      stops: <double>[0.4, 1.0],
    );
    canvas.drawRect(
      rect,
      Paint()..shader = gradient.createShader(rect),
    );
    List<Rect> _playerTracks = [];
    _playerTracks.add(rect);
    trackCalculationListener(_playerTracks);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}

```
</details>

**Comment by ratedRA** [comment]
@huycozy Well I need that value in the postFrameCallbacks and using the above workaround also doesnt resolve that issue. But, I did a separate workaround for this wherein I am generating that value in the current screen (mainMenu) and sending it to next (game) screen with pushNamed args and it seems to have resolved it.

**Comment by exaby73** [comment]
## Triage report

I am able to reproduce the issue with the code sample provided by @ratedRA. Below is a code sample that I've modified to put in `print` statements instead. From the output of these print statements, looks like `paint` is being called after `build` and `addPostFrameCallback`. Labeling the issue for further insights from the team

<details>
<summary>Code Sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      routes: {
        '/mainMenu': (context) => const MainMenu(),
        '/game': (context) => const MyHomePage(title: 'game'),
      },
      initialRoute: '/mainMenu',
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late List<Rect> _playerTracks;

  @override
  void initState() {
    super.initState();
    print('initState called');

    WidgetsBinding.instance.addPostFrameCallback((_) {
      // print(_playerTracks.length);
      print('addPostFrameCallback called');
    });
  }

  @override
  Widget build(BuildContext context) {
    print('build called');
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Container(
              color: Colors.white,
              margin: const EdgeInsets.all(20),
              child: AspectRatio(
                aspectRatio: 1,
                child: SizedBox.expand(
                  child: CustomPaint(
                    painter: RectanglePainter(
                        trackCalculationListener: (playerTracks) {
                      print('trackCalculationListener called');
                      _playerTracks = playerTracks;
                    }),
                  ),
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}

class MainMenu extends StatefulWidget {
  static String route = '/mainMenu';
  const MainMenu({super.key});

  @override
  State<MainMenu> createState() => _MainMenuState();
}

class _MainMenuState extends State<MainMenu> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Container(
          height: 200.0,
          color: Colors.white,
          child: ElevatedButton(
            onPressed: () {
              Navigator.pushNamed(context, '/game');
            },
            child: const Text('play game'),
          ),
        ),
      ),
    );
  }
}

class RectanglePainter extends CustomPainter {
  Function(List<Rect>) trackCalculationListener;

  RectanglePainter({required this.trackCalculationListener});

  @override
  void paint(Canvas canvas, Size size) {
    print('paint called');

    final Rect rect = Offset.zero & size;
    const RadialGradient gradient = RadialGradient(
      center: Alignment(0.7, -0.6),
      radius: 0.2,
      colors: <Color>[Color(0xFFFFFF00), Color(0xFF0099FF)],
      stops: <double>[0.4, 1.0],
    );
    canvas.drawRect(
      rect,
      Paint()..shader = gradient.createShader(rect),
    );
    List<Rect> playerTracks = [];
    playerTracks.add(rect);
    trackCalculationListener(playerTracks);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}
```

</details>

<details>
<summary>Output</summary>

```console
flutter: initState called
flutter: build called
flutter: addPostFrameCallback called
flutter: paint called
flutter: trackCalculationListener called
```

</details>

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[!] Flutter (Channel master, 3.7.0-17.0.pre.23, on macOS 13.0.1 22A400 darwin-arm64, locale en-US)
    • Flutter version 3.7.0-17.0.pre.23 on channel master at /Users/nabeelparkar/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.18.6/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nabeelparkar/fvm/versions/master. Consider adding /Users/nabeelparkar/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 32b1ea3b58 (59 minutes ago), 2023-01-11 01:46:08 -0500
    • Engine revision 8d7beac822
    • Dart version 3.0.0 (build 3.0.0-109.0.dev)
    • DevTools version 2.20.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-33, build-tools 33.0.0
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.3.1)
    • IntelliJ at /Users/nabeelparkar/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 71.3.6
    • Dart plugin version 223.8214.16

[✓] IntelliJ IDEA Ultimate Edition (version 2022.3.1)
    • IntelliJ at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/223.8214.52/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.74.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (4 available)
    • iPhone 14 Pro (mobile)     • A75A00DB-CD34-4A0D-BFD9-2DA753E5A5EE • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-2 (simulator)
    • iPhone 14 Pro Max (mobile) • 6BC61E88-4DE0-4BCB-8599-DA5294269146 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-2 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 13.0.1 22A400 darwin-arm64
    • Chrome (web)               • chrome                               • web-javascript • Brave Browser 108.1.46.153

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

</details>

**Comment by goderbauer** [comment]
In general, there is no guarantee that a painter callback will get called before the end of the frame. It is possible that a widget is just not getting painted in a given frame. So, a design that relies on a painter callback getting called before the end of a frame is just flawed.
